bindings/bindings.py

```diff
@@ -0,0 +1,593 @@
+        )
+        for code, resp in self.responses.items():
+            out += f"\n       {code} = {resp}"
+        out += "\n    }"
+        return out
+
+    def gen_def(self) -> Code:
+        # Function name.
+        out = [f"def {self.method}_{self.name}("]
+
+        # Function parameters.
+        out += ['    do_request: Request,']
+        if self.params:
+            out += ["    *,"]
+
+        required = sorted(p for p in self.params if self.params[p].required)
+        optional = sorted(p for p in self.params if not self.params[p].required)
+        for name in required + optional:
+            out += [self.params[name].gen_function_param()]
+
+        # Function return type.
+        # (simplifying assumptions; if broken we need more logic)
+        responses = {**self.responses}
+        default = responses.pop("default")
+        assert isinstance(default, Ref) and default.name == "runtimeError", (self.name, default)
+
+        if len(responses) == 1:
+            returntype = next(iter(responses.values()))
+            returntypestr = returntype.annotation()
+        else:
+            returntypes = set(r.annotation(prequoted=True) for r in responses.values())
+            returntypestr = '"Union[' + ", ".join(sorted(returntypes)) + ']"'
+        assert len(responses) == 1, (self.name, responses)
+
+        out += [f') -> {returntypestr}:']
+
+        # Function body.
+        path_params = sorted(p for p in self.params if self.params[p].where == "path")
+        body_params = sorted(p for p in self.params if self.params[p].where == "body")
+        query_params = sorted(p for p in self.params if self.params[p].where == "query")
+
+        pathstr = f'"{self.path}"'
+        if path_params:
+            # Happily, we can just generate an f-string based on the path swagger gives us.
+            pathstr = "f" + pathstr
+
+        if query_params:
+            out += ["    _params = {"]
+            for p in query_params:
+                out += [f'        "{self.params[p].serialized_name}": {self.params[p].name},']
+            out += ["    }"]
+        else:
+            out += ["    _params = None"]
+
+        if "body" in self.params:
+            bodystr = self.params["body"].dump()
+        else:
+            bodystr = "None"
+        out += ["    _req = do_request("]
+        out += [f"        \"{self.method.upper()}\","]
+        out += [f"        {pathstr},"]
+        out += ["        _params,"]
+        out += [f"        {bodystr},"]
+        out += ["        None,"]
+        out += ["        None,"]
+        out += ["        None,"]
+        out += ["    )"]
+        for expect, returntype in responses.items():
+            out += [f"    if _req.status_code == {expect}:"]
+            if returntype.isnone():
+                out += ['        return']
+            else:
+                out += [f'        return {returntype.load("_req.json()")}']
+        # XXX: raise something else here
```

Comment:
maybe add a small custom `class APIHttpError(Exception):` in the prefix, and throw it here?
It'd also be very helpful for debugging and logging to include the entire response object on the exception, and not just the status code

Reply:
sounds good this would still be dependent on the passed in http request to not raise any expectations (or do if that's what the user of the bindings wants) unless we opt to wrap and catch those as well